### PR TITLE
APEXMALHAR-2218-Creation of RegexSplitter operator

### PR DESCRIPTION
**Pull request by venkateshkottapalli**
APEXMALHAR-2218-Creation of RegexSplitter operator

@ashwinchandrap Please review.

### CODE REVIEW

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+ * <br>
+ * <b>Properties</b> <br>
+ * <b>splitRegexPattern</b>:Regex pattern as a string<br>
+ *
+ * @displayName RegexSplitter
+ * @category Parsers
+ * @tags pojo parser regex logs server
+ * @since 3.6.0
+ */
+public class RegexSplitter extends CsvParser
```
Comment (shubham-pathak22) [inline contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java]: Any specific reason to extend from CsvParser ?  I feel RegexSplitter is not a "special case" of Csv Parser. It should extend from Parser instead of CsvParser.

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+{
+  @NotNull
+  private  String splitRegexPattern;
```
Comment (shubham-pathak22) [inline contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java]: Javadoc. In general please add JavaDocs for field members and getters and setters.

**Comment by shubham-pathak22** [comment]
Would adding another port just emitting tokens after the split be useful ?  That way this operator could also be used as a tokenizer

File: contrib/src/test/java/com/datatorrent/contrib/parser/RegexSplitterTest.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  @Test
+  public void testNullInput() throws JSONException
+  {
+    regex.beginWindow(0);
+    regex.in.process(null);
+    regex.endWindow();
+    Assert.assertEquals(0, pojoPort.collectedTuples.size());
+    Assert.assertEquals(1, error.collectedTuples.size());
+  }
+
```
Comment (shubham-pathak22) [inline contrib/src/test/java/com/datatorrent/contrib/parser/RegexSplitterTest.java]: You may add tests to verify values of counters ( autometrics ) as well.

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java (hunk 1 above)
Comment (shubham-pathak22) [reply contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java]: Mark the operator as @InterfaceStability.Evolving

Comment (venkateshkottapalli) [reply contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java]: Agree. Will make the changes suggested.

File: contrib/src/main/java/com/datatorrent/contrib/parser/CsvParser.java [hunk 4]
```diff
@@ -65,7 +65,7 @@
  * <b>err</b>:tuples that do not confine to schema are emitted on this port as
  * KeyValPair<String,String><br>
  * Key being the tuple and Val being the reason.
- * 
+ *
```
Comment (shubham-pathak22) [inline contrib/src/main/java/com/datatorrent/contrib/parser/CsvParser.java]: Please undo these changes. Spaces have been introduced.

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+      if (err.isConnected()) {
+        err.emit(new KeyValPair<String, String>(null, "Blank/null tuple"));
+      }
+      errorTupleCount++;
+      return;
+    }
+    String incomingString = new String(tuple);
+    if (StringUtils.isBlank(incomingString)) {
+      if (err.isConnected()) {
+        err.emit(new KeyValPair<String, String>(incomingString, "Blank/header tuple"));
```
Comment (shubham-pathak22) [inline contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java]: Error message could just be "Blank tuple" as the check is for blank string

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+  /**
+   * Set the Regex Pattern expected for the incoming tuple
+   *
+   * @param splitRegexPattern
+   */
+  public void setSplitRegexPattern(String splitRegexPattern)
+  {
+    this.splitRegexPattern = splitRegexPattern;
+  }
+
```
Comment (shubham-pathak22) [inline contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java]: add getters for both schema and splitRegexPattern

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+  @Override
+  public void processTuple(byte[] tuple)
+  {
+    if (tuple == null) {
+      if (err.isConnected()) {
+        err.emit(new KeyValPair<String, String>(null, "Blank/null tuple"));
+      }
+      errorTupleCount++;
```
Comment (shubham-pathak22) [inline contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java]: Where are the counters being reset ? You might need to call super.beginWindow().
In addition to it please add a test to check whether counters are being reset for each window. You may refer to 
https://github.com/apache/apex-malhar/blob/master/contrib/src/test/java/com/datatorrent/contrib/parser/JsonParserTest.java

**Comment by shubham-pathak22** [comment]
There are build failures because of missing headers. Good practice is to run this command before pushing in the changes
mvn apache-rat:check verify -Dlicense.skip=false -Dtravis=true -Dtest.log.console.threshold=INFO -Pall-modules -B -V

File: contrib/src/test/java/com/datatorrent/contrib/parser/RegexSplitterTest.java [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    public void setDate(Date date)
+    {
+      this.date = date;
+    }
+
+  }
+
+  private static final Logger logger = LoggerFactory.getLogger(RegexSplitterTest.class);
```
Comment (shubham-pathak22) [inline contrib/src/test/java/com/datatorrent/contrib/parser/RegexSplitterTest.java]: You may remove the logger since it is not being used

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java (hunk 7 above)
Comment (venkateshkottapalli) [reply contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java]: Counters are reset in the Parser beginWindow(). It gets called since we don't have an implementation in RegexSplitter.

I will add test cases to check that like you suggested.

File: contrib/src/test/java/com/datatorrent/contrib/parser/RegexSplitterTest.java (hunk 8 above)
Comment (venkateshkottapalli) [reply contrib/src/test/java/com/datatorrent/contrib/parser/RegexSplitterTest.java]: Sure.

**Comment by shubham-pathak22** [comment]
@venkateshkottapalli  The changes look good to me. Will merge in a day if there are no further comments.

**Comment by venkateshkottapalli** [comment]
Travis build failing.. Will close and reopen.

**Comment by venkateshkottapalli** [comment]
APEXMALHAR-2218-Creation of RegexSplitter operator - Reopen

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+ * <b>Properties</b> <br>
+ * <b>splitRegexPattern</b>:Regex pattern as a string<br>
+ *
+ * @displayName RegexSplitter
+ * @category Parsers
+ * @tags pojo parser regex logs server
+ * @since 3.6.0
+ */
+@InterfaceStability.Evolving
+public class RegexSplitter extends Parser<byte[], KeyValPair<String, String>>
```
Comment (bhupeshchawda) [inline contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java]: Can this be called something like RegexParser?
RegexSplitter does not intuitively seem like some parser. It looks like a type of `FilterOperator`, but one that operates on raw data.

Comment (venkateshkottapalli) [reply contrib/src/main/java/com/datatorrent/contrib/parser/RegexSplitter.java]: @bhupeshchawda Agreed, RegexParser looks more appropriate.

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+      return;
+    }
+    try {
+      if (out.isConnected() && clazz != null) {
+        Matcher matcher = pattern.matcher(incomingString);
+        boolean patternMatched = false;
+        Constructor<?> ctor = clazz.getConstructor();
+        Object object = ctor.newInstance();
+
+        while (matcher.find()) {
```
Comment (devtagare) [inline contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java]: Why is this not a if condition

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+          for (int i = 1; i <= matcher.groupCount(); i++) {
+            if (delimitedParserSchema.getFields().get(i - 1).getType() == DelimitedSchema.FieldType.DATE) {
+              DateTimeConverter dtConverter = new DateConverter();
+              dtConverter.setPattern((String)delimitedParserSchema.getFields().get(i - 1).getConstraints().get(DelimitedSchema.DATE_FORMAT));
+              ConvertUtils.register(dtConverter, Date.class);
+            }
+            BeanUtils.setProperty(object, delimitedParserSchema.getFields().get(i - 1).getName(), matcher.group(i));
```
Comment (devtagare) [inline contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java]: Use PojoUtils instead.BeanUtils uses reflection

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+          }
+          patternMatched = true;
+        }
+        if (!patternMatched) {
+          throw new ConversionException("The incoming tuple do not match with the Regex pattern defined.");
```
Comment (devtagare) [inline contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java]: Please dont throw an exception here emit to error port

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+import com.datatorrent.lib.parser.Parser;
+import com.datatorrent.lib.util.KeyValPair;
+
+/**
+ * Operator parsers the tuple based on regex pattern and populates POJO matching the user defined schema <br>
+ * This operator expects the upstream operator to send every line in the file as a byte array.
+ * splitRegexPattern contains the regex pattern of lines in the file <br>
+ * Schema is specified in a json format as per {@link DelimitedSchema} that
+ * contains field information and constraints for each field.<br>
+ * Schema field names should match with the POJO variable names<br>
```
Comment (devtagare) [inline contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java:52]: Why is this condition enforced.You can use https://github.com/apache/apex-malhar/blob/master/library/src/main/java/com/datatorrent/lib/util/FieldInfo.java to specify input to output mappings

File: contrib/src/test/java/com/datatorrent/contrib/parser/RegexParserTest.java [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+import org.junit.runner.Description;
+
+import com.datatorrent.lib.appdata.schemas.SchemaUtils;
+import com.datatorrent.lib.testbench.CollectorTestSink;
+import com.datatorrent.lib.util.KeyValPair;
+
+public class RegexParserTest
+{
+  RegexParser regex = new RegexParser();
+  private CollectorTestSink<Object> error = new CollectorTestSink<Object>();
```
Comment (devtagare) [inline contrib/src/test/java/com/datatorrent/contrib/parser/RegexParserTest.java:39]: Pleas add an application level test

File: contrib/src/test/resources/RegexSplitterschema.json [hunk 15]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (devtagare) [inline contrib/src/test/resources/RegexSplitterschema.json:1]: Does this operator need a CSV like schema definition ? How is this operator using a separator and quouteChar for parsing inputs ?

Comment (venkateshkottapalli) [reply contrib/src/test/resources/RegexSplitterschema.json:1]: @devtagare It doesn't need the delimiters in schema definition, will remove them from the test schema.

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java (hunk 13 above)
Comment (venkateshkottapalli) [inline contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java]: @devtagare Thanks for suggesting this and letting me know. Adding an additional mapping also makes sense but this may not be needed as the user has control over both schema and POJO.  If the schema has say 50 or more fields, it gets tedious to set these fields, we need not make it mandatory though by adding additional checks. Let me know your thoughts and did we see any issues with this type of constraint before?

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java (hunk 10 above)
Comment (venkateshkottapalli) [reply contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java]: Both if and while serve the same purpose here, but using if is appropriate, will change it.

File: contrib/src/test/java/com/datatorrent/contrib/parser/RegexParserTest.java (hunk 14 above)
Comment (venkateshkottapalli) [reply contrib/src/test/java/com/datatorrent/contrib/parser/RegexParserTest.java:39]: @devtagare any reason for an application test in Unit tests?

**Comment by sandeshh** [comment]
What is the status of this PR?

**Comment by venkateshkottapalli** [comment]
I will incorporate the changes in a couple of days Shubham.. I am creating
an example to see the bottlenecks.

-Venky.

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java (hunk 11 above)
Comment (venkateshkottapalli) [reply contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java]: PojoUtils doesn't support Date fields. Using PojoUtils for the rest and Reflection for Date can be tried, if that is feasible solution I will track it as a different issue. At present, a single instance of an operator is able to process 32K records per second with 2 strings, 1 long and 1 Date field. The test case has the String and the regex pattern example.

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java (hunk 13 above)
Comment (venkateshkottapalli) [reply contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java:52]: There is no input to output mapping here as in the case of Json.  Schema field names  define the structure of the String but will not be present in data. Please look at the test case for sample data.

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java (hunk 10 above)
Comment (shubham-pathak22) [reply contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java]: Change this to "if" as suggested by @devtagare

**Comment by shubham-pathak22** [comment]
@venkateshkottapalli Looks good to me. If there are no further comments and once the changes are incorporated,  please rebase and make sure the build is successful. Will merge the PR then.

File: contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java (hunk 10 above)
Comment (venkateshkottapalli) [reply contrib/src/main/java/com/datatorrent/contrib/parser/RegexParser.java]: @shubham-pathak22 Done with the change.

**Comment by venkateshkottapalli** [comment]
Build failing.. will close and reopen!
